tests/components/http/test_forwarded.py

```diff
@@ -0,0 +1,487 @@
+    async_setup_forwarded(app, [])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/",
+        headers={X_FORWARDED_FOR: "255.255.255.255", X_FORWARDED_HOST: "example.com"},
+    )
+
+    assert resp.status == 200
+
+
+async def test_x_forwarded_host_with_trusted_proxy(aiohttp_client):
+    """Test that we get the host header if proxy is trusted."""
+
+    async def handler(request):
+        assert request.host == "example.com"
+        assert request.scheme == "http"
+        assert not request.secure
+        assert request.remote == "255.255.255.255"
+
+        return web.Response()
+
+    app = web.Application()
+    app.router.add_get("/", handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/",
+        headers={X_FORWARDED_FOR: "255.255.255.255", X_FORWARDED_HOST: "example.com"},
+    )
+
+    assert resp.status == 200
+
+
+async def test_x_forwarded_host_not_processed_without_for(aiohttp_client):
+    """Test that host header isn't processed without a for header."""
+
+    async def handler(request):
+        url = mock_api_client.make_url("/")
+        assert request.host == f"{url.host}:{url.port}"
+        assert request.scheme == "http"
+        assert not request.secure
+        assert request.remote == "127.0.0.1"
+
+        return web.Response()
+
+    app = web.Application()
+    app.router.add_get("/", handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get("/", headers={X_FORWARDED_HOST: "example.com"})
+
+    assert resp.status == 200
+
+
+async def test_x_forwarded_host_with_multiple_headers(aiohttp_client, caplog):
+    """Test that we get a HTTP 400 bad request with multiple headers."""
+    app = web.Application()
+    app.router.add_get("/", mock_handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/",
+        headers=[
+            (X_FORWARDED_FOR, "222.222.222.222"),
+            (X_FORWARDED_HOST, "example.com"),
+            (X_FORWARDED_HOST, "example.spoof"),
+        ],
+    )
+
+    assert resp.status == 400
+    assert "Too many headers for X-Forwarded-Host" in caplog.text
+
+
+async def test_x_forwarded_host_with_empty_header(aiohttp_client, caplog):
+    """Test that we get a HTTP 400 bad request with empty host value."""
+    app = web.Application()
+    app.router.add_get("/", mock_handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/", headers={X_FORWARDED_FOR: "222.222.222.222", X_FORWARDED_HOST: ""}
+    )
+
+    assert resp.status == 400
+    assert "Empty value received in X-Forward-Host header" in caplog.text
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)